### PR TITLE
Clean up cross bootstrapping, take 2

### PR DESCRIPTION
**Pull request by Ericson2314**
## Description of changes

Take 2 of #320852, without the stdenv change that necessitated the revert in #321435.

Now following up from https://github.com/NixOS/nixpkgs/pull/320852, which got the other part of #320852.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Ericson2314** [comment]
Part of this I would like to backport, adjusting the history accordingly.

**Comment by Ericson2314** [comment]
Pulled out the stuff that can be backported in #321538.

File: pkgs/build-support/bintools-wrapper/default.nix [hunk 1]
```diff
@@ -10,15 +10,15 @@
 , stdenvNoCC
 , runtimeShell
 , bintools ? null, libc ? null, coreutils ? null, gnugrep ? null
-, netbsd ? null, netbsdCross ? null
+, netbsd ? null
```
Comment (AndersonTorres) [inline pkgs/build-support/bintools-wrapper/default.nix:13]: Not to be overzealous, but  that excess of `? null` is dangerous.

Comment (Ericson2314) [reply pkgs/build-support/bintools-wrapper/default.nix:13]: Sure, but it's been this way for sake of bootstrapping for years. I hope after I find time to redo the GCC bootstrapping to dramatically simplify all this stuff :).

**Comment by Ericson2314** [comment]
Draft because I want to do it after the the other things that I will also backport, to avoid conflict.

**Review by philiptaron — COMMENTED** [review]
I'll give this the thorough review and merge if rebased, John (@Ericson2314).

File: pkgs/stdenv/cross/default.nix [hunk 2]
```diff
@@ -41,25 +41,44 @@ in lib.init bootStages ++ [
...
+
+      cc = null;
+      hasCC = false;
+
+      extraNativeBuildInputs = old.extraNativeBuildInputs
+        ++ lib.optionals
+             (hostPlatform.isLinux && !buildPlatform.isLinux)
+             [ buildPackages.patchelf ]
+        ++ lib.optional
+             (let f = p: !p.isx86 || builtins.elem p.libc [ "musl" "wasilibc" "relibc" ] || p.isiOS || p.isGenode;
```
Comment (philiptaron) [inline pkgs/stdenv/cross/default.nix:63]: What a strange assortment...

**Review by philiptaron — APPROVED** [review]
All right, building the normal things appears to work. Building strange things like `nix-build pkgs/top-level/release-cross.nix -A riscv64.rustc` gets substantially further than it does on master.

File: pkgs/build-support/bintools-wrapper/default.nix [hunk 3]
```diff
@@ -10,15 +10,15 @@
 , stdenvNoCC
 , runtimeShell
 , bintools ? null, libc ? null, coreutils ? null, gnugrep ? null
-, netbsd ? null, netbsdCross ? null
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply pkgs/build-support/bintools-wrapper/default.nix:13]: Nah the NetBSD stuff is still too new and experimental to worry about that.

File: pkgs/stdenv/cross/default.nix (hunk 2 above)
Comment (Ericson2314) [reply pkgs/stdenv/cross/default.nix:63]: (it was pulled from below)

**Comment by Ericson2314** [comment]
Accidentally reintroduced a darwin stdenv rebuild, should go away now.

File: pkgs/stdenv/cross/default.nix (hunk 2 above)
Comment (philiptaron) [reply pkgs/stdenv/cross/default.nix:63]: Yeah, I see that. It's, um, quite the list.

**Review by philiptaron — APPROVED** [review]

**Pull request merged**